File: leetcode/python/2406.divide-intervals-into-minimum-number-of-groups.py
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        intervals.sort()
        heap = [] # minheap
        for i in intervals:
            if heap and heap[0] < i[0]:
                heapq.heappop(heap)
            heapq.heappush(heap, i[1])
        
        return len(heap)

File: leetcode/rust-leetcode/src/2406.divide-intervals-into-minimum-number-of-groups.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;
impl Solution {
    pub fn min_groups(mut intervals: Vec<Vec<i32>>) -> i32 {
        let mut heap:BinaryHeap<Reverse<i32>> = BinaryHeap::new();
        intervals.sort();//sort_by_key(|i| i[0]);
        // println!("{:?}", intervals);
        for i in intervals.iter() {
            if !heap.is_empty() && i[0] > heap.peek().unwrap().0 {
                heap.pop();
            }
            heap.push(Reverse(i[1]));
        }
        heap.len() as _
    }
}

